File: src/test/java/no/nav/bidrag/commons/web/test/HttpHeaderTestRestTemplateTest.java
package no.nav.bidrag.commons.web.test;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertAll;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

import java.util.List;
import java.util.Objects;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentCaptor;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.springframework.boot.test.web.client.TestRestTemplate;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;


@DisplayName("HttpHeaderTestRestTemplateTest")
class HttpHeaderTestRestTemplateTest {

  private HttpHeaderTestRestTemplate httpHeaderTestRestTemplate;
  @Mock
  private TestRestTemplate testRestTemplateMock;

  @BeforeEach
  void initClassToTestWithCustomHeader() {
    MockitoAnnotations.initMocks(this);
    httpHeaderTestRestTemplate = new HttpHeaderTestRestTemplate(testRestTemplateMock);
    httpHeaderTestRestTemplate.add("X-Custom", () -> "my custom header");
  }

  @Test
  @DisplayName("skal legge ved custom header verdi ved bruk")
  void skalLeggeVedCustomHeaderVerdiVedBruk() {
    httpHeaderTestRestTemplate.exchange("somewhere", null, new HttpEntity<>(null, null), Object.class);

    ArgumentCaptor<HttpEntity> entityCaptor = ArgumentCaptor.forClass(HttpEntity.class);
    verify(testRestTemplateMock).exchange(anyString(), any(), entityCaptor.capture(), eq(Object.class));

    assertThat(entityCaptor.getValue()).isNotNull();

    var httpEntity = entityCaptor.getValue();

    assertAll(
        () -> assertThat(httpEntity.getHeaders()).as("headers").hasSize(1),
        () -> assertThat(Objects.requireNonNull(httpEntity.getHeaders().get("X-Custom")).get(0))
            .as("headerValue").contains("my custom header")
    );
  }

  @Test
  @DisplayName("skal ikke overskrive eksisterende headere")
  void skalIkkeOverskriveEksisterendeHeadere() {
    HttpHeaders headers = new HttpHeaders();
    headers.add("something", "completely different");
    httpHeaderTestRestTemplate.exchange("somewhere", null, new HttpEntity<>(null, headers), Object.class);

    ArgumentCaptor<HttpEntity> entityCaptor = ArgumentCaptor.forClass(HttpEntity.class);
    verify(testRestTemplateMock).exchange(anyString(), any(), entityCaptor.capture(), eq(Object.class));

    assertThat(entityCaptor.getValue()).isNotNull();
    assertThat(entityCaptor.getValue().getHeaders()).as("headers").hasSize(2);
  }

  @Test
  @DisplayName("skal inititalisere ny HttpEntity når argument er null")
  void skalInitialisereNyHttpEntityNaarArgumentErNull() {
    ParameterizedTypeReference<List<Object>> typeReference = new ParameterizedTypeReference<>() {
    };

    httpHeaderTestRestTemplate.exchange("somewhere", null, null, typeReference);

    ArgumentCaptor<HttpEntity> entityCaptor = ArgumentCaptor.forClass(HttpEntity.class);
    verify(testRestTemplateMock).exchange(anyString(), any(), entityCaptor.capture(), eq(typeReference));

    assertThat(entityCaptor.getValue()).isNotNull();

    var httpEntity = entityCaptor.getValue();

    assertAll(
        () -> assertThat(httpEntity.getHeaders()).as("headers").hasSize(1),
        () -> assertThat(Objects.requireNonNull(httpEntity.getHeaders().get("X-Custom")).get(0))
            .as("headerValue").contains("my custom header")
    );
  }

  @Test
  @DisplayName("skal legge til HEADER som bare er gyldig for èn HttpEntity callback")
  void skalLeggeTilHeaderSomErGyldigForEnHttpEntityCallback() {
    ParameterizedTypeReference<List<Object>> typeReference = new ParameterizedTypeReference<>() {
    };

    httpHeaderTestRestTemplate.addHeaderForSingleHttpEntityCallback("X-OnlyOnce", "WithValue");

    assertAll(
        () -> {
          httpHeaderTestRestTemplate.exchange("somewhere", null, null, typeReference);

          ArgumentCaptor<HttpEntity> entityCaptor = ArgumentCaptor.forClass(HttpEntity.class);
          verify(testRestTemplateMock).exchange(anyString(), any(), entityCaptor.capture(), eq(typeReference));

          assertThat(entityCaptor.getValue()).isNotNull();

          var httpEntity = entityCaptor.getValue();

          assertAll(
              () -> assertThat(httpEntity.getHeaders()).as("headers").hasSize(2),
              () -> assertThat(Objects.requireNonNull(httpEntity.getHeaders().get("X-Custom")).get(0))
                  .as("Custom header value").contains("my custom header"),
              () -> assertThat(Objects.requireNonNull(httpEntity.getHeaders().get("X-OnlyOnce")).get(0))
                  .as("header for one callback").contains("WithValue")
          );
        }, () -> {
          httpHeaderTestRestTemplate.exchange("somewhere", null, null, typeReference);

          ArgumentCaptor<HttpEntity> entityCaptor = ArgumentCaptor.forClass(HttpEntity.class);
          verify(testRestTemplateMock, times(2)).exchange(anyString(), any(), entityCaptor.capture(), eq(typeReference));

          assertThat(entityCaptor.getValue()).isNotNull();

          var httpEntity = entityCaptor.getValue();

          assertAll(
              () -> assertThat(httpEntity.getHeaders()).as("headers").hasSize(1),
              () -> assertThat(Objects.requireNonNull(httpEntity.getHeaders().get("X-Custom")).get(0))
                  .as("headerValue").contains("my custom header")
          );

        }
    );
  }

  @Nested
  @DisplayName("... as secured testRestTemplate")
  class Secured {

    @BeforeEach
    void addSecurityHeader() {
      httpHeaderTestRestTemplate.add(HttpHeaders.AUTHORIZATION, () -> "Bearer secured");
    }


    @Test
    @DisplayName("skal legge på OIDC token før exchange-kall med TestRestTemplate")
    void skalLeggePaOidcToken() {
      httpHeaderTestRestTemplate.exchange("somewhere", null, new HttpEntity<>(null, null), Object.class);

      ArgumentCaptor<HttpEntity> entityCaptor = ArgumentCaptor.forClass(HttpEntity.class);
      verify(testRestTemplateMock).exchange(anyString(), any(), entityCaptor.capture(), eq(Object.class));

      assertThat(entityCaptor.getValue()).isNotNull();

      var httpEntity = entityCaptor.getValue();

      assertAll(
          () -> assertThat(httpEntity.getHeaders()).as("headers").hasSize(2), // custom and security header
          () -> assertThat(Objects.requireNonNull(httpEntity.getHeaders().get(HttpHeaders.AUTHORIZATION)).get(0))
              .as("token").contains("Bearer ")
      );
    }

    @Test
    @SuppressWarnings("unchecked")
    @DisplayName("skal inititalisere ny HttpEntity når argument er null")
    void skalInitialisereNyHttpEntityNaarArgumentErNull() {
      httpHeaderTestRestTemplate.exchange("somewhere", null, null, new ParameterizedTypeReference<List<Object>>() {
      });

      ArgumentCaptor<HttpEntity> entityCaptor = ArgumentCaptor.forClass(HttpEntity.class);
      verify(testRestTemplateMock).exchange(anyString(), any(), entityCaptor.capture(), any(ParameterizedTypeReference.class));

      assertThat(entityCaptor.getValue()).isNotNull();

      var httpEntity = entityCaptor.getValue();

      assertAll(
          () -> assertThat(httpEntity.getHeaders()).as("headers").hasSize(2), // custom and security header
          () -> assertThat(Objects.requireNonNull(httpEntity.getHeaders().get(HttpHeaders.AUTHORIZATION)).get(0))
              .as("token").contains("Bearer ")
      );
    }
  }
}

File: src/test/kotlin/no/nav/bidrag/commons/web/test/HttpHeaderTestRestTemplateTest.kt
package no.nav.bidrag.commons.web.test

import io.kotest.matchers.maps.shouldHaveSize
import io.kotest.matchers.shouldNotBe
import io.kotest.matchers.string.shouldContain
import io.mockk.CapturingSlot
import io.mockk.mockk
import io.mockk.verify
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.springframework.boot.test.web.client.TestRestTemplate
import org.springframework.core.ParameterizedTypeReference
import org.springframework.http.HttpEntity
import org.springframework.http.HttpHeaders
import java.util.*

internal class HttpHeaderTestRestTemplateTest {

    private lateinit var httpHeaderTestRestTemplate: HttpHeaderTestRestTemplate

    private val testRestTemplateMock: TestRestTemplate = mockk(relaxed = true)

    @BeforeEach
    fun initClassToTestWithCustomHeader() {
        httpHeaderTestRestTemplate = HttpHeaderTestRestTemplate(testRestTemplateMock)
        httpHeaderTestRestTemplate.add("X-Custom") { "my custom header" }
    }

    @Test
    fun `exchange skal legge ved custom header verdi ved bruk`() {
        val slot = CapturingSlot<HttpEntity<*>>()

        httpHeaderTestRestTemplate.exchange("somewhere", null, HttpEntity<Any?>(null, null), Any::class.java)

        verify {
            testRestTemplateMock.exchange(any<String>(), any(), capture(slot), any<Class<*>>())
        }
        slot.captured shouldNotBe null
        val httpEntity = slot.captured
        httpEntity.headers shouldHaveSize 1
        httpEntity.headers["X-Custom"]!!.first() shouldContain "my custom header"
    }

    @Test
    fun `exchange skal ikke overskrive eksisterende headere`() {
        val headers = HttpHeaders().apply { add("something", "completely different") }
        val slot = CapturingSlot<HttpEntity<*>>()

        httpHeaderTestRestTemplate.exchange("somewhere", null, HttpEntity<Any?>(null, headers), Any::class.java)

        verify {
            testRestTemplateMock.exchange(any<String>(), any(), capture(slot), any<Class<*>>())
        }
        slot.captured shouldNotBe null
        slot.captured.headers shouldHaveSize 2
    }

    @Test
    fun `exchange skal inititalisere ny HttpEntity når argument er null`() {
        val typeReference: ParameterizedTypeReference<List<Any>> = ParameterizedTypeReference.forType(Any::class.java)
        val slot = CapturingSlot<HttpEntity<*>>()

        httpHeaderTestRestTemplate.exchange("somewhere", null, null, typeReference)

        verify {
            testRestTemplateMock.exchange(any<String>(), any(), capture(slot), eq(typeReference))
        }
        slot.captured shouldNotBe null
        val httpEntity = slot.captured
        httpEntity.headers shouldHaveSize 1
        httpEntity.headers["X-Custom"]!!.first() shouldContain "my custom header"
    }

    @Test
    fun `exchange skal legge til HEADER som bare er gyldig for èn HttpEntity callback`() {
        val typeReference: ParameterizedTypeReference<List<Any>> = ParameterizedTypeReference.forType(Any::class.java)

        httpHeaderTestRestTemplate.addHeaderForSingleHttpEntityCallback("X-OnlyOnce", "WithValue")
        httpHeaderTestRestTemplate.exchange("somewhere", null, null, typeReference)
        httpHeaderTestRestTemplate.exchange("somewhere", null, null, typeReference)

        val slot = mutableListOf<HttpEntity<*>>()
        verify(exactly = 2) {
            testRestTemplateMock.exchange(any<String>(), any(), capture(slot), eq(typeReference))
        }
        slot.first() shouldNotBe null
        val firstHttpEntity = slot.first()
        firstHttpEntity.headers shouldHaveSize 2
        firstHttpEntity.headers["X-Custom"]!!.first() shouldContain "my custom header"
        firstHttpEntity.headers["X-OnlyOnce"]!!.first() shouldContain "WithValue"
        slot[1] shouldNotBe null
        val httpEntity = slot[1]
        httpEntity.headers shouldHaveSize 1
        httpEntity.headers["X-Custom"]!!.first() shouldContain "my custom header"

    }

    @Nested
    @DisplayName("... as secured testRestTemplate")
    internal inner class Secured {
        @BeforeEach
        fun addSecurityHeader() {
            httpHeaderTestRestTemplate.add(HttpHeaders.AUTHORIZATION) { "Bearer secured" }
        }

        @Test
        fun `exchange skal legge på OIDC token før exchange-kall med TestRestTemplate`() {
            httpHeaderTestRestTemplate.exchange("somewhere", null, HttpEntity<Any?>(null, null), Any::class.java)
            val entityCaptor = CapturingSlot<HttpEntity<*>>()
            verify {
                testRestTemplateMock.exchange(any<String>(), any(), capture(entityCaptor), eq(Any::class.java))
            }
            entityCaptor.captured shouldNotBe null
            val httpEntity = entityCaptor.captured
            httpEntity.headers shouldHaveSize 2   // custom and security header
            httpEntity.headers[HttpHeaders.AUTHORIZATION]!!.first() shouldContain "Bearer "
        }

        @Test
        fun `exchange skal inititalisere ny HttpEntity når argument er null`() {
            httpHeaderTestRestTemplate.exchange<List<Any>>("somewhere", null, null, ParameterizedTypeReference.forType(Any::class.java))
            val entityCaptor = CapturingSlot<HttpEntity<*>>()
            verify {
                testRestTemplateMock.exchange(
                    any<String>(), any(), capture(entityCaptor), any<ParameterizedTypeReference<List<Any>>>()
                )
            }

            entityCaptor.captured shouldNotBe null
            val httpEntity = entityCaptor.captured

            httpEntity.headers shouldHaveSize 2   // custom and security header
            httpEntity.headers[HttpHeaders.AUTHORIZATION]!!.first() shouldContain "Bearer "
        }
    }
}